sem3/homework1/LazyInterface/LazyInterfaceTests/LazyInterfaceTests.cs

```
@@ -0,0 +1,173 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace LazyInterface.Tests
+{
+    [TestClass]
+    public class LazyInterfaceTests
+    {
+        // one thread tests
+
+        [TestMethod]
+        public void OneThreadSimpleMethodTest()
+        {
+            Func<float> simple = () => 0.1f;
+            var lazy = LazyFactory<float>.CreateOneThreading(simple);
+            Assert.AreEqual(0.1f, lazy.Get());
+        }
+
+        [TestMethod]
+        public void OneThreadPowTest()
+        {
+            Func<double> addition = () =>
+            {
+                return Math.Pow(2.0, 30.0);
+            };
+            var lazy = LazyFactory<double>.CreateOneThreading(addition);
+            Assert.AreEqual(lazy.Get(), 1073741824.0);
+        }
+
+        [TestMethod]
+        public void OneThreadNullTest()
+        {
+            Func<string> empty = () => null;
+            var lazy = LazyFactory<string>.CreateOneThreading(empty);
+            Assert.IsNull(lazy.Get());
+        }
+
+        [TestMethod]
+        public void OneThreadNullMethodTest()
+        {
+            var lazy = LazyFactory<string>.CreateOneThreading(null);
+            Assert.ThrowsException<NullReferenceException>(lazy.Get);
```

Нуу, я бы ещё в фабрике ArgumentNullException бросал, а не невнятное NullReferenceException при попытке сделать Get. Fail fast же.